nixos/lib/test-driver/test-driver.py

```diff
@@ -738,10 +759,19 @@ def create_socket(path: str) -> socket.socket:
         self.monitor, _ = self.monitor_socket.accept()
         self.shell, _ = self.shell_socket.accept()
 
+        # Store last serial console lines for use
+        # of wait_for_console_text
+        self.last_lines: Queue = Queue(maxsize=1000)
+
         def process_serial_output() -> None:
             for _line in self.process.stdout:
                 # Ignore undecodable bytes that may occur in boot menus
                 line = _line.decode(errors="ignore").replace("\r", "").rstrip()
+                try:
+                    self.last_lines.put(line, block=False)
+                except queue.Full:
```

Comment:
i would rather see first that we really have the problem that some test fails because it goes OOM or gets too slow processing serial output, before we begin dealing with silently failing tests because we drop serial lines without further notice because a small 1000-line buffer overruns.

Reply:
Sorry, I don't what you're asking here. Do you want this operation to be blocking or are you proposing to increase the buffer size (as above)?

Reply:
i think that it would be a good idea to:

- make the buffer size infinite (get that with empty Queue constructor args)
- make the `put` calls blocking, assuming the queue never runs full
- revisit the question if it is really necessary to model the notion of a 100-lines "screen", because it makes a difference for the caller if the "screen buffer" has 1, 10, or 100 screens stored currently. there is no way to get to the "last" screen. wouldn't it suffice to give `wait_for_console_text` semantics that are like "does the serial output match my regex since boot/i called this last time"?

Reply:
Ok, I removed the size limit and stored the lines received since `wait_for_console_text` is called in a `StringIO`.
This way we efficiently append lines into the buffer and can match multiline regexes.

Reply:
ping @tfc

Reply:
LGTM!